### PR TITLE
Improve wrapper cradle errors

### PR DESCRIPTION
**Pull request by hasufell**
@jhrcek 

wrt https://github.com/haskell/haskell-language-server/pull/2675#issuecomment-1039053247

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2711"></a>

### CODE REVIEW

**Comment by hasufell** [comment]
@jhrcek build via:

```
ghcup compile hls -r https://github.com/hasufell/haskell-language-server.git -g improve-wrapper-error --ghc 8.10.7
```

**Comment by jhrcek** [comment]
Thank you @hasufell , you're really fast!

I'm getting this error:
```
Failed to get project GHC executable path:CradleError {cradleErrorDependencies = [], cradleErrorExitCode = ExitFailure 1, cradleErrorStderr = ["Error when calling stack exec ghc -- -v0 -package-env=- -e putStr =<< System.Environment.getExecutablePath","","\n<interactive>:0:1: error: Variable not in scope: putStr\n\n<interactive>:0:8: error:\n    Variable not in scope:\n      (=<<) :: t0 -> GHC.Types.IO GHC.IO.FilePath -> t\n"]}
```

**Comment by hasufell** [comment]
> Thank you @hasufell , you're really fast!
> 
> I'm getting this error:
> 
> ```
> Failed to get project GHC executable path:CradleError {cradleErrorDependencies = [], cradleErrorExitCode = ExitFailure 1, cradleErrorStderr = ["Error when calling stack exec ghc -- -v0 -package-env=- -e putStr =<< System.Environment.getExecutablePath","","\n<interactive>:0:1: error: Variable not in scope: putStr\n\n<interactive>:0:8: error:\n    Variable not in scope:\n      (=<<) :: t0 -> GHC.Types.IO GHC.IO.FilePath -> t\n"]}
> ```

Uh... can you run

```
stack exec ghc -- -v0 -package-env=- -e putStr =<< System.Environment.getExecutablePath"
```

inside your project? What if you run it outside of your project? Is `NoImplicitPrelude` activated?

**Comment by hasufell** [comment]
Yeah, pretty sure it's `NoImplicitPrelude`:

```
stack exec ghc -- -XNoImplicitPrelude -v0 -package-env=- -e 'putStr =<< System.Environment.getExecutablePath'

<interactive>:0:1: error: Variable not in scope: putStr

<interactive>:0:8: error:
    Variable not in scope:
      (=<<) :: t0 -> GHC.Types.IO GHC.IO.FilePath -> t
```

**Review by pepeiborra — APPROVED** [review]

**Comment by jhrcek** [comment]
Outside of any project it works fine
```
~> stack exec ghc -- -v0 -package-env=- -e "putStr =<< System.Environment.getExecutablePath"
/home/jhrcek/.stack/programs/x86_64-linux/ghc-tinfo6-8.10.7/lib/ghc-8.10.7/bin/ghc
```

within our work codebase it doesn't:

```bash
~/Devel/github.com/Holmusk/PI/backend> stack exec ghc -- -v0 -package-env=- -e "putStr =<< System.Environment.getExecutablePath"

<interactive>:0:1: error: Variable not in scope: putStr

<interactive>:0:8: error:
    Variable not in scope:
      (=<<) :: t0 -> GHC.Types.IO GHC.IO.FilePath -> t
```

Indeed we are using NoImplicitPrelude :smile:

**Comment by hasufell** [comment]
Ok, I'll push a fix to this PR.

**Comment by hasufell** [comment]
@jhrcek please try again

**Comment by jhrcek** [comment]
I wonder if it woulndn't be maybe easier to do
```
$ stack exec ghc -- -v0  -e "System.Environment.getExecutablePath"
"/home/jhrcek/.stack/programs/x86_64-linux/ghc-tinfo6-8.10.7/lib/ghc-8.10.7/bin/ghc"
```

and then strip the quotes?

Or alternatively use `System.IO.putStr`?

**Comment by hasufell** [comment]
`stack exec ghc -- -XNoImplicitPrelude -v0 -package-env=- -e 'do e <- System.Environment.getExecutablePath ; System.IO.putStr e'`

**Comment by jhrcek** [comment]
That works ok
```
stack exec ghc -- -XNoImplicitPrelude -v0 -package-env=- -e 'do e <- System.Environment.getExecutablePath ; System.IO.putStr e'
/home/jhrcek/.stack/programs/x86_64-linux/ghc-tinfo6-8.10.7/lib/ghc-8.10.7/bin/ghc
```

**Review by jhrcek — APPROVED** [review]
Thank you very much for quick response. Works for me :+1:

File: exe/Wrapper.hs [hunk 1]
```diff
@@ -1,5 +1,6 @@
 {-# LANGUAGE RecordWildCards #-}
 {-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE LambdaCase #-}
```
Comment (jhrcek) [inline exe/Wrapper.hs:3]: Seems unused after refactoring
```suggestion
```

File: exe/Wrapper.hs [hunk 2]
```diff
@@ -114,13 +114,21 @@ launchHaskellLanguageServer parsedArgs = do
       callProcess e args
 #else
       let Cradle { cradleOptsProg = CradleAction { runGhcCmd } } = cradle
-      (CradleSuccess ghcBinary) <- fmap trim <$> runGhcCmd ["-v0", "-package-env=-", "-e", "putStr =<< System.Environment.getExecutablePath"]
-      (CradleSuccess libdir) <- HieBios.getRuntimeGhcLibDir cradle
+      ghcBinary <- (fmap trim <$> runGhcCmd ["-v0", "-package-env=-", "-e", "do e <- System.Environment.getExecutablePath ; System.IO.putStr e"])
```
Comment (michaelpj) [inline exe/Wrapper.hs:117]: Worth a comment, very non-obvious that you're doing it this way to be compatible with `NoImplicitPrelude`.

Also, this probably won't work with `RebindableSyntax` still, but surely nobody sets *that* globally...

**Review by michaelpj — APPROVED** [review]

**Pull request merged**